src/client/keycode.cpp

```diff
@@ -221,122 +223,153 @@ static const struct table_key table[] = {
+	return invalid_key;
 }
 
-static const table_key &lookup_keychar(wchar_t Char)
+static const table_key &lookup_keyname(std::string_view name)
 {
+	if (name.empty())
+		return invalid_key;
+
 	for (const auto &table_key : table) {
-		if (table_key.Char == Char)
+		if (table_key.Name == name)
 			return table_key;
 	}
 
-	std::ostringstream os;
-	os << "<Char " << hex_encode((char*) &Char, sizeof(wchar_t)) << ">";
-	throw UnknownKeycode(os.str().c_str());
+	auto wname = utf8_to_wide(name);
+	if (wname.empty())
+		return invalid_key;
+	return lookup_keychar(wname[0]);
 }
 
-KeyPress::KeyPress(const char *name)
+static const table_key &lookup_scancode(const u32 scancode)
 {
-	if (strlen(name) == 0) {
-		Key = irr::KEY_KEY_CODES_COUNT;
-		Char = L'\0';
-		m_name = "";
-		return;
-	}
+	auto key = RenderingEngine::get_raw_device()->getKeyFromScancode(scancode);
+	return std::holds_alternative<EKEY_CODE>(key) ?
+		lookup_keykey(std::get<irr::EKEY_CODE>(key)) :
+		lookup_keychar(std::get<wchar_t>(key));
+}
 
-	if (strlen(name) <= 4) {
-		// Lookup by resulting character
-		int chars_read = mbtowc(&Char, name, 1);
-		FATAL_ERROR_IF(chars_read != 1, "Unexpected multibyte character");
-		try {
-			auto &k = lookup_keychar(Char);
-			m_name = k.Name;
-			Key = k.Key;
-			return;
-		} catch (UnknownKeycode &e) {};
-	} else {
-		// Lookup by name
-		m_name = name;
-		try {
-			auto &k = lookup_keyname(name);
-			Key = k.Key;
-			Char = k.Char;
-			return;
-		} catch (UnknownKeycode &e) {};
-	}
+static const table_key &lookup_scancode(const std::variant<u32, irr::EKEY_CODE> &scancode)
+{
+	return std::holds_alternative<irr::EKEY_CODE>(scancode) ?
+		lookup_keykey(std::get<irr::EKEY_CODE>(scancode)) :
+		lookup_scancode(std::get<u32>(scancode));
+}
 
-	// It's not a known key, complain and try to do something
-	Key = irr::KEY_KEY_CODES_COUNT;
-	int chars_read = mbtowc(&Char, name, 1);
-	FATAL_ERROR_IF(chars_read != 1, "Unexpected multibyte character");
-	m_name = "";
-	warningstream << "KeyPress: Unknown key '" << name
-		<< "', falling back to first char." << std::endl;
+KeyPress::KeyPress(std::string_view name)
+{
+	if (loadFromScancode(name))
+		return;
+	const auto &key = lookup_keyname(name);
+	Keycode keycode(key.Key, key.Char);
+	scancode = RenderingEngine::get_raw_device()->getScancodeFromKey(keycode);
 }
 
-KeyPress::KeyPress(const irr::SEvent::SKeyInput &in, bool prefer_character)
+KeyPress::KeyPress(const irr::SEvent::SKeyInput &in)
 {
-	if (prefer_character)
-		Key = irr::KEY_KEY_CODES_COUNT;
+#if USE_SDL2
+	if (in.SystemKeyCode)
+		scancode.emplace<u32>(in.SystemKeyCode);
 	else
-		Key = in.Key;
-	Char = in.Char;
-
-	try {
-		if (valid_kcode(Key))
-			m_name = lookup_keykey(Key).Name;
-		else
-			m_name = lookup_keychar(Char).Name;
-	} catch (UnknownKeycode &e) {
-		m_name.clear();
-	};
+		scancode.emplace<irr::EKEY_CODE>(in.Key);
+#else
+	Keycode keycode(in.Key, in.Char);
+	scancode = RenderingEngine::get_raw_device()->getScancodeFromKey(keycode);
```

Comment:
should include a sanity_check before dereferencing the device

Reply:
Like this: https://github.com/luanti-org/luanti/pull/14964/commits/106cb91b9665df101fde8b87b324c6f2754edf6c ?

Reply:
sure. I think it should even raise a warning if the device is not initialized yet. since that means the result won't be correct.

Reply:
290498b37